src/routing_table/mod.rs

```diff
@@ -694,4 +701,132 @@ mod tests {
+        let mut rng: XorShiftRng = SeedableRng::from_seed([1315, 30, 61894, 315]);
+        let our_name = rng.next_u32();
+        let mut table = RoutingTable::new(our_name, 8);
+        let mut unknown_distant_name = None;
+
+        for _ in 0..1000 {
+            let new_name = rng.next_u32();
+            // Try to add new_name. We double-check the output to test this too.
+            match table.add(new_name) {
+                Err(Error::AlreadyExists) => {
+                    assert!(table.iter().any(|u| *u == new_name));
+                    // skip
+                }
+                Err(Error::PeerNameUnsuitable) => {
+                    assert!(table.groups.keys().all(|p| !p.matches(&new_name)));
+                    // We should get a few of these. Save one for tests, but otherwise ignore.
+                    unknown_distant_name = Some(new_name);
+                }
+                Err(_) => {
+                    assert!(false); // no other errors should be possible
+                }
+                Ok(true) => {
+                    assert!(table.our_group_prefix.matches(&new_name));
+                    assert!(unwrap!(table.groups.get(&table.our_group_prefix)).len() >=
+                            table.min_split_size() * 2);
+                    let our_prefix = *table.our_group_prefix();
+                    let _ = table.split(our_prefix);
+                    assert!(unwrap!(table.groups.get(&table.our_group_prefix)).len() >=
+                            table.min_split_size());
+                }
+                Ok(false) => {
+                    assert!(table.iter().any(|u| *u == new_name));
+                    if table.is_in_our_group(&new_name) {
+                        continue;   // add() already checked for necessary split
+                    }
+
+                    // Not a split event for our group, but might be for a different group.
+                    let group_prefix = table.find_group_prefix(&new_name)
+                        .expect("get group added to");
+                    let (group_len, new_group_size) = {
+                        let group = table.groups.get(&group_prefix).expect("get group from prefix");
+                        // Count size of group after an arbitrary split (note that there is only
+                        // one split possible; the arbitrariness is just which half we choose here).
+                        (group.len(),
+                         group.iter()
+                            .filter(|name| new_name.common_prefix(name) > group_prefix.bit_count())
+                            .count())
+                    };
+                    let min_size = table.min_split_size();
+                    if new_group_size >= min_size && group_len - new_group_size >= min_size {
+                        let _ = table.split(group_prefix);  // do the split
+                    }
+                }
+            }
+        }
+
+        let unknown_close_name;
```

Comment:
Maybe this shouldn't be called `close`, as it's explicitly not in our own group, whereas `close_name` below is.
Maybe `unknown_neighbouring_name`? :grimacing: